grug/malachite/src/actors/host/host.rs

```diff
@@ -0,0 +1,560 @@
+
+            reply_to.send(LocallyProposedValue::new(
+                value.height,
+                value.round,
+                value.value,
+            ))?;
+
+            return Ok(());
+        }
+
+        // Crate proposal parts
+        let block = {
+            // Take txs from mempool
+            info!("Taking txs from mempool");
+            let txs = self
+                .mempool
+                .call(|reply| MempoolMsg::Take { reply }, None)
+                .await?
+                .success_or(anyhow::anyhow!("Failed to take txs from mempool"))?;
+
+            // Call prepare_proposal
+            info!("Calling prepare_proposal");
+            let txs = state.prepare_proposal(txs);
+
+            let pre_block = PreBlock::new(
+                height,
+                self.address,
+                round,
+                Timestamp::from_nanos(
+                    std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)
+                        .unwrap()
+                        .as_nanos(),
+                ),
+                txs,
+            );
+
+            info!("Calling finalize_block");
+            let block_hash = state.finalize_block(&pre_block)?;
+
+            pre_block.into_block(block_hash)
+        };
+
+        let proposal_data = block.as_proposal_data();
+
+        let stream_id = state.stream_id();
+
+        // Store undecided proposal
+        let proposed_value = ProposedValue {
+            height,
+            round,
+            valid_round: Round::Nil,
+            proposer: self.address,
+            value: <ctx!(Value)>::new(proposal_data.block.block_hash()),
+            validity: Validity::Valid,
+        };
+
+        UNDECIDED_PROPOSALS.save(
+            state,
+            (*height, round.as_i64(), proposal_data.block.block_hash()),
+            &proposed_value,
+        )?;
+
+        // Store undecided block
+        UNDECIDED_BLOCK.save(state, proposal_data.block.block_hash(), &block)?;
+
+        let value = LocallyProposedValue::new(
+            height,
+            round,
+            <ctx!(Value)>::new(proposal_data.block.block_hash()),
+        );
+
+        // Stream parts to consensus
+        self.stream_data(stream_id, proposal_data, true).await?;
+
+        // Return the proposed value
+        reply_to.send(value)?;
+
+        Ok(())
+    }
+
+    #[tracing::instrument("received_proposal_part", skip_all)]
+    async fn received_proposal_part(
+        &self,
+        state: &mut State,
+        part: StreamMessage<ctx!(ProposalPart)>,
+        reply_to: RpcReplyPort<ProposedValue<Context>>,
+    ) -> ActorResult<()> {
+        // Since we are using a single part proposal, we only need data
+        let Some(ProposalData { block, valid_round }) = part.content.into_data() else {
+            return Ok(());
+        };
+
+        info!(height = %block.height, round = %block.round, "All parts received");
+
```

Comment:
Check that the received block timestamp is after the previous block timestamp. If not the proposal should be marked `Invalid` and sent to consensus. 
You can check the comet implementation [here](https://github.com/cometbft/cometbft/blob/bd1bc3bb6392a973d7b60fd0b19da6e71ca09b47/types/proposal.go#L85-L107).
It is a bit more complex as it takes into account clock drifts and message delays which you may want to add at some point.